credentials/java/com/google/auth/Credentials.java

```diff
@@ -36,6 +37,34 @@
   }
 
   /**
+   * Get the current request metadata without blocking.
+   *
+   * <p>This should be called by the transport layer on each request, and the data should be
+   * populated in headers or other context. The implementation can either call the callback inline,
+   * or from the given executor asynchronously. Either way it should <strong>never block</strong> in
+   * this method.
+   *
+   * <p>The default implementation will just call {@link #getRequestMetadata(URI)} then the callback
+   * from the given executor.
+   *
+   * <p>The convention for handling binary data is for the key in the returned map to end with
+   * {@code "-bin"} and for the corresponding values to be base64 encoded.
+   */
+  public void getRequestMetadata(final URI uri, Executor executor, final Callback callback) {
+    executor.execute(new Runnable() {
+        @Override
+        public void run() {
+          try {
+            Map<String, List<String>> result = getRequestMetadata(uri);
+            callback.onSuccess(result);
```

Comment:
If the callback throws, should that call onFailure? Hmm... tough call. I think I'd prefer it propagating to the `executor` simply so that each implementation isn't supposed to worry about the issue.